File: backend/app/services/rag.py
import json
import os
from dataclasses import dataclass, asdict
from pathlib import Path

import numpy as np
import structlog

from app.services.llm import generate_embedding, generate_embeddings

logger = structlog.get_logger()

CACHE_DIR = Path(os.getcwd()) / ".cache"
EMBEDDINGS_CACHE = CACHE_DIR / "guideline-embeddings.json"


@dataclass
class EmbeddedChunk:
    id: str
    text: str
    guideline_id: str
    title: str
    source: str
    embedding: list[float]


@dataclass
class RAGResult:
    text: str
    guideline_id: str
    title: str
    source: str
    score: float


_embedded_chunks: list[EmbeddedChunk] | None = None


def cosine_similarity(a: list[float], b: list[float]) -> float:
    a_arr = np.array(a)
    b_arr = np.array(b)
    dot = np.dot(a_arr, b_arr)
    norm = np.linalg.norm(a_arr) * np.linalg.norm(b_arr)
    if norm == 0:
        return 0.0
    return float(dot / norm)


def _load_cached_embeddings() -> list[EmbeddedChunk] | None:
    try:
        if EMBEDDINGS_CACHE.exists():
            data = json.loads(EMBEDDINGS_CACHE.read_text())
            if isinstance(data, list) and len(data) > 0 and "embedding" in data[0]:
                return [
                    EmbeddedChunk(
                        id=d["id"],
                        text=d["text"],
                        guideline_id=d.get("guidelineId", d.get("guideline_id", "")),
                        title=d["title"],
                        source=d["source"],
                        embedding=d["embedding"],
                    )
                    for d in data
                ]
    except Exception:
        pass
    return None


def _save_cached_embeddings(chunks: list[EmbeddedChunk]) -> None:
    try:
        CACHE_DIR.mkdir(parents=True, exist_ok=True)
        # Save with camelCase keys for compatibility with the TS cache
        data = [
            {
                "id": c.id,
                "text": c.text,
                "guidelineId": c.guideline_id,
                "title": c.title,
                "source": c.source,
                "embedding": c.embedding,
            }
            for c in chunks
        ]
        EMBEDDINGS_CACHE.write_text(json.dumps(data))
    except Exception as e:
        logger.error("failed_to_save_embeddings_cache", error=str(e))


async def initialize_rag() -> None:
    global _embedded_chunks

    if _embedded_chunks is not None:
        return

    cached = _load_cached_embeddings()
    if cached:
        _embedded_chunks = cached
        logger.info("loaded_cached_embeddings", count=len(cached))
        return

    logger.info("generating_guideline_embeddings")
    from app.services.guidelines import get_guideline_chunks

    chunks = get_guideline_chunks()
    texts = [c.text for c in chunks]

    all_embeddings: list[list[float]] = []
    for i in range(0, len(texts), 10):
        batch = texts[i : i + 10]
        batch_embeddings = await generate_embeddings(batch)
        all_embeddings.extend(batch_embeddings)

    _embedded_chunks = [
        EmbeddedChunk(
            id=chunk.id,
            text=chunk.text,
            guideline_id=chunk.guideline_id,
            title=chunk.title,
            source=chunk.source,
            embedding=all_embeddings[i],
        )
        for i, chunk in enumerate(chunks)
    ]

    _save_cached_embeddings(_embedded_chunks)
    logger.info("generated_and_cached_embeddings", count=len(_embedded_chunks))


async def search_rag(query: str, top_k: int = 5) -> list[RAGResult]:
    global _embedded_chunks

    if not _embedded_chunks:
        await initialize_rag()

    if not _embedded_chunks:
        raise RuntimeError("RAG pipeline not initialized — no guideline embeddings available")

    query_embedding = await generate_embedding(query)

    scored = [
        RAGResult(
            text=chunk.text,
            guideline_id=chunk.guideline_id,
            title=chunk.title,
            source=chunk.source,
            score=cosine_similarity(query_embedding, chunk.embedding),
        )
        for chunk in _embedded_chunks
    ]

    scored.sort(key=lambda r: r.score, reverse=True)

    seen: set[str] = set()
    deduped: list[RAGResult] = []
    for result in scored:
        if result.guideline_id not in seen:
            seen.add(result.guideline_id)
            deduped.append(result)
        if len(deduped) >= top_k:
            break

    return deduped


def search_rag_fallback(query: str, top_k: int = 5) -> list[RAGResult]:
    import re
    from app.services.guidelines import get_guideline_chunks

    chunks = get_guideline_chunks()
    terms = [t for t in query.lower().split() if len(t) > 2]

    scored: list[RAGResult] = []
    for chunk in chunks:
        text = chunk.text.lower()
        score = 0.0
        for term in terms:
            escaped = re.escape(term)
            matches = len(re.findall(escaped, text))
            score += matches
        score = score / max(len(terms), 1)
        scored.append(
            RAGResult(
                text=chunk.text,
                guideline_id=chunk.guideline_id,
                title=chunk.title,
                source=chunk.source,
                score=score,
            )
        )

    scored.sort(key=lambda r: r.score, reverse=True)

    seen: set[str] = set()
    deduped: list[RAGResult] = []
    for result in scored:
        if result.score > 0 and result.guideline_id not in seen:
            seen.add(result.guideline_id)
            deduped.append(result)
        if len(deduped) >= top_k:
            break

    return deduped

File: src/lib/rag.ts
/**
 * RAG (Retrieval-Augmented Generation) Pipeline
 * 
 * Uses Gemini text-embedding-004 for embeddings and cosine similarity
 * for retrieval over the medical guidelines knowledge base.
 * Embeddings are cached in a JSON file on disk for fast restarts.
 */

import { generateEmbedding, generateEmbeddings } from './gemini';
import { getGuidelineChunks } from './guidelines';
import * as fs from 'fs';
import * as path from 'path';

const CACHE_DIR = path.join(process.cwd(), '.cache');
const EMBEDDINGS_CACHE = path.join(CACHE_DIR, 'guideline-embeddings.json');

interface EmbeddedChunk {
    id: string;
    text: string;
    guidelineId: string;
    title: string;
    source: string;
    embedding: number[];
}

interface RAGResult {
    text: string;
    guidelineId: string;
    title: string;
    source: string;
    score: number;
}

let embeddedChunks: EmbeddedChunk[] | null = null;

function cosineSimilarity(a: number[], b: number[]): number {
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < a.length; i++) {
        dotProduct += a[i] * b[i];
        normA += a[i] * a[i];
        normB += b[i] * b[i];
    }
    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
}

function loadCachedEmbeddings(): EmbeddedChunk[] | null {
    try {
        if (fs.existsSync(EMBEDDINGS_CACHE)) {
            const data = JSON.parse(fs.readFileSync(EMBEDDINGS_CACHE, 'utf-8'));
            if (Array.isArray(data) && data.length > 0 && data[0].embedding) {
                return data;
            }
        }
    } catch {
        // Cache corrupted, will regenerate
    }
    return null;
}

function saveCachedEmbeddings(chunks: EmbeddedChunk[]): void {
    try {
        if (!fs.existsSync(CACHE_DIR)) {
            fs.mkdirSync(CACHE_DIR, { recursive: true });
        }
        fs.writeFileSync(EMBEDDINGS_CACHE, JSON.stringify(chunks));
    } catch (err) {
        console.error('Failed to save embeddings cache:', err);
    }
}

export async function initializeRAG(): Promise<void> {
    // Check if already initialized
    if (embeddedChunks) return;

    const currentChunks = getGuidelineChunks();
    const currentChunkIds = new Set(currentChunks.map(c => c.id));

    // Try loading from cache
    const cached = loadCachedEmbeddings();
    if (cached && cached.length === currentChunks.length && cached.every(c => currentChunkIds.has(c.id))) {
        embeddedChunks = cached;
        console.log(`Loaded ${cached.length} cached guideline embeddings`);
        return;
    } else if (cached) {
        console.log('Guideline corpus changed; regenerating guideline embeddings');
    }

    // Generate fresh embeddings
    console.log('Generating guideline embeddings (first run)...');
    const texts = currentChunks.map(c => c.text);

    // Batch embed in groups of 10
    const allEmbeddings: number[][] = [];
    for (let i = 0; i < texts.length; i += 10) {
        const batch = texts.slice(i, i + 10);
        const batchEmbeddings = await generateEmbeddings(batch);
        allEmbeddings.push(...batchEmbeddings);
    }

    embeddedChunks = currentChunks.map((chunk, i) => ({
        ...chunk,
        embedding: allEmbeddings[i],
    }));

    // Cache for next startup
    saveCachedEmbeddings(embeddedChunks);
    console.log(`Generated and cached ${embeddedChunks.length} guideline embeddings`);
}

export async function searchRAG(query: string, topK: number = 5): Promise<RAGResult[]> {
    if (!embeddedChunks || embeddedChunks.length === 0) {
        await initializeRAG();
    }

    if (!embeddedChunks || embeddedChunks.length === 0) {
        throw new Error('RAG pipeline not initialized — no guideline embeddings available');
    }

    // Embed the query
    const queryEmbedding = await generateEmbedding(query);

    // Compute similarity scores
    const scored = embeddedChunks.map(chunk => ({
        text: chunk.text,
        guidelineId: chunk.guidelineId,
        title: chunk.title,
        source: chunk.source,
        score: cosineSimilarity(queryEmbedding, chunk.embedding),
    }));

    // Sort by score descending and return top-k
    scored.sort((a, b) => b.score - a.score);

    // Deduplicate by guidelineId, keeping the highest-scoring chunk per guideline
    const seen = new Set<string>();
    const deduped: RAGResult[] = [];
    for (const result of scored) {
        if (!seen.has(result.guidelineId)) {
            seen.add(result.guidelineId);
            deduped.push(result);
        }
        if (deduped.length >= topK) break;
    }

    return deduped;
}

/**
 * Fallback keyword-based search when API key is not configured
 */
export function searchRAGFallback(query: string, topK: number = 5): RAGResult[] {
    const chunks = getGuidelineChunks();
    const terms = query.toLowerCase().split(/\s+/).filter(t => t.length > 2);

    const scored = chunks.map(chunk => {
        const text = chunk.text.toLowerCase();
        let score = 0;
        for (const term of terms) {
            const escaped = term.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
            const matches = (text.match(new RegExp(escaped, 'g')) || []).length;
            score += matches;
        }
        // Normalize
        score = score / Math.max(terms.length, 1);
        return {
            text: chunk.text,
            guidelineId: chunk.guidelineId,
            title: chunk.title,
            source: chunk.source,
            score,
        };
    });

    scored.sort((a, b) => b.score - a.score);

    const seen = new Set<string>();
    const deduped: RAGResult[] = [];
    for (const result of scored) {
        if (result.score > 0 && !seen.has(result.guidelineId)) {
            seen.add(result.guidelineId);
            deduped.push(result);
        }
        if (deduped.length >= topK) break;
    }

    return deduped;
}
